src/NHibernate/AdoNet/GenericBatchingBatcher.cs

```diff
@@ -176,9 +176,11 @@ public void Append(DbParameterCollection parameters)
 						_batcher.CurrentCommandParameterTypes);
 					UpdateCommandParameters(command, parameters);
 					_batchCommand.CommandText += $"{_batcher.StatementTerminator}{command.CommandText}";
-					foreach (DbParameter parameter in command.Parameters)
+					while (command.Parameters.Count > 0)
 					{
-						_batchCommand.Parameters.Add(CopyParameter(_batchCommand, parameter));
+						var pram = command.Parameters[0];
+						command.Parameters.RemoveAt(0);
```

Comment:
Removing fist element is relatively slow for lists (as array is recopied on each call). It would be much more effective to remove last element instead. As It seems order of parameters is not important.

https://dotnetfiddle.net/ZKRkkb (it shows 4x difference, but on my machine in Release mode it's closer to 10x)

Reply:
Unfortunately, order of parameters _is_ important for some dialects

Reply:
@bahusoid thanks for pointing it out! I redesigned the logic in order to call `GenerateCommand` only once and removed the code for removing parameters. After that I tested the difference and I was shocked when I saw the results. The new approach is 10x faster than the old one, [here](https://github.com/maca88/nhibernate-core/commit/7656d846b97b70e85237ce0eee1b1d37793dc1be) is how I setup the test. 
Old BatchingCommandSet took avg: `4959ms`
New BatchingCommandSet took avg: `435ms`